src/redis-cli.c

```diff
@@ -7726,40 +7739,62 @@ unsigned long long sendSync(redisContext *c, char *out_eof) {
         fprintf(stderr, "SYNC with master failed: %s\n", buf);
         exit(1);
     }
+
+    /* Handling PSYNC responses. */
+    if (!strncmp(buf, "+FULLRESYNC", 11) ||
+        !strncmp(buf, "+CONTINUE", 9))
+    {
+        int sync_partial = !strncmp(buf, "+CONTINUE", 9);
+        fprintf(stderr, "PSYNC replied %s\n", buf);
+        p = buf;
+        while(1) {
+            nread = readConn(c,p,1);
+            if (nread <= 0) {
+                fprintf(stderr,"Error reading bulk length while PSYNCing\n");
+                exit(1);
+            }
+            if (*p == '\n' && p != buf) break;
+            if (*p != '\n') p++;
+        }
+        *p = '\0';
+
+        if (sync_partial) return 0;
+    }
+
     if (strncmp(buf+1,"EOF:",4) == 0 && strlen(buf+5) >= RDB_EOF_MARK_SIZE) {
         memcpy(out_eof, buf+5, RDB_EOF_MARK_SIZE);
         return 0;
     }
     return strtoull(buf+1,NULL,10);
 }
 
-static void slaveMode(void) {
+static void slaveMode(int is_psync) {
     static char eofmark[RDB_EOF_MARK_SIZE];
     static char lastbytes[RDB_EOF_MARK_SIZE];
     static int usemark = 0;
-    unsigned long long payload = sendSync(context,eofmark);
+    unsigned long long payload = sendSync(context,eofmark,!is_psync);
```

Comment:
> could it be that the user typed SYNC in the prompt, and this condition is wrong?

we have two call to the `slaveMode`, one is the using redis-cli --replica, the others one is the command in the prompt.  in the prompt one, we will use `slaveMode(!strcasecmp(command, "psync"));`, so then type SYNC in the prompt, the condition is also ok.

> i.e. the argument to slaveMode() is if we want to use partial_sync or not.
and the argument to sendSync() is if we want to send any command at all. right?

yes, and yes

Reply:
so if the user types `SYNC` in the prompt:
we'll call `slaveMode(0)`, and that will call `sendSync(1)`

and if the user types `PSYNC` in the prompt
we'll call `slaveMode(1)`, and that will call `sendSync(0)`

don't we need to call `sendSync` with 0 in both cases?

Reply:
> don't we need to call sendSync with 0 in both cases?

yes, in the SYNC case, we will send SYNC command twice (the untable branch code way)
and syncCommand ignore the second SYNC because `if (c->flags & CLIENT_SLAVE) return;`
i am changing it, now they both call sendSync with 0. (change the argument in slaveMode, from is_psync to send_sync)